src/config/passport.conf.js

```diff
@@ -118,50 +124,83 @@ export default (passport) => {
       return done(null, false, { loginMessage : 'Invalid password length.' });
     }
 
-    User.findOne({
-      'local.email' : email.toLowerCase()
-    }, (err, user) => {
-      if (err) return done(err);
-
-      // If no user is found, return a message
-      if (!user) {
-        return done(
-          null,
-          false,
-          { loginMessage : 'That user was not found. ' +
-          'Please enter valid user credentials.' }
-        );
-      }
-
-      // If the user is found but the password is incorrect
-      if (!user.validPassword(password)) {
-        return done(null, false, { loginMessage: 'Invalid password entered.' });
-      }
-
-      // Check account suspension
-      if (user.suspendedUntil && user.suspendedUntil > new Date()) {
-        return done(null, false, { loginMessage: 'This account is suspended.' });
-      }
-
-      // If the user's email address is not verified yet
-      if (!user.verified && (!user.role || user.role !== 'ROLE_ADMIN')) {
-        return done(null, false, { loginMessage: 'Please confirm your email address.'});
-      }
-
-      // Otherwise all is well; return successful user
-
-      // Generate hash for Gravatar if not present
-      if (!user.gravatarHash) {
-        user.gravatarHash = md5(user.local.email);
-
-        // Save the new user
-        user.save(err => {
-          if (err) throw err;
+    User.findOne(
+      (email.indexOf('@') !== -1) ? { 'local.email' : email.toLowerCase() } : { 'username' : email },
+      async (err, user) => {
+        if (err) return done(err);
+
+        // If no user is found, return a message
+        if (!user) {
+          return done(
+            null,
+            false,
+            { loginMessage : 'That user was not found. ' +
+            'Please enter valid user credentials.' }
+          );
+        }
+
+        // If the user is found but the password is incorrect
+        if (!user.validPassword(password)) {
+          return done(null, false, { loginMessage: 'Invalid password entered.' });
+        }
+
+        // Check account suspension
+        if (user.suspendedUntil && user.suspendedUntil > new Date()) {
+          return done(null, false, { loginMessage: 'This account is suspended.' });
+        }
+
+        // If the user's email address is not verified yet
+        if (!user.verified && (!user.role || user.role !== 'ROLE_ADMIN')) {
+          return done(null, false, { loginMessage: 'Please confirm your email address.'});
+        }
+
+        // Check if a username is set and if not request one
+        if (!user.username) {
+          if (!req.body.username || reservedUsernames.contains(req.body.username)) {
```

Comment:
Whoops, I meant [includes](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/includes). Will fix it.